corehq/apps/users/validation.py

```diff
@@ -0,0 +1,47 @@
+from django.core.exceptions import ValidationError
+from django.core.validators import validate_email
+
+from corehq.apps.users.dbaccessors import user_exists
+from corehq.apps.users.exceptions import (
+    InvalidDomainException,
+    InvalidUsernameException,
+    ReservedUsernameException,
+    UsernameAlreadyExists,
+)
+from corehq.apps.users.util import format_username
+from corehq.apps.users.views.mobile import BAD_MOBILE_USERNAME_REGEX
+
+
+def validate_mobile_username(username, domain):
+    if username is None:
+        raise InvalidUsernameException
+    if domain is None:
+        raise InvalidDomainException
+
+    _check_for_reserved_usernames(username)
+    username_as_email = format_username(username, domain)
+    _ensure_valid_username(username_as_email)
```

Comment:
I see the order here has changed. Is it possible that the 'username_as_email' won't be a valid email address? For example if the domain name is somehow invalid as an actual domain?

Previously we were using 'example.com' to validate the username, just wondering if there was a reason for doing it that way.

Reply:
Good point. Interestingly enough, in `normalize_username` we format the username with the domain first, and then call the same `validate_email` method:
https://github.com/dimagi/commcare-hq/blob/2968eeac63baed5dc1f088c907d5d00cdb8db06c/corehq/apps/users/util.py#L71-L73

That was added in 2011, whereas the "@example.com" version in `check_username` was added in 2018, so that portion of this check seems safe. 

However the `BAD_MOBILE_USERNAME_REGEX` is only used in check_usernames. I tested this locally by creating a domain "my%domain%", which is saved as "my-domain" to be used in URLs, and it looks like that url version is what is used when accessing a domain's name. I think this is safe to keep as is, but let me know if you have any concerns.

Reply:
Thanks, that makes sense. Re the domain name - I agree that should be safe. I think the user input value is saved in `domain.hr_name` (human readable name).